File: neuron-di-it/src/test/java/global/tranquillity/neuron/di/it/GuiceContextIT.java
package global.tranquillity.neuron.di.it;

import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import com.google.inject.Injector;
import global.tranquillity.neuron.di.guice.GuiceContext;
import org.junit.Test;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import static com.google.inject.name.Names.named;
import static org.hamcrest.Matchers.instanceOf;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.*;

public class GuiceContextIT {

    @Test
    public void testGuiceContextDSL() {
        assertInjector(new GuiceContext()
                .injector()
                    .module()
                        .bindConstant().annotatedWith(named("i")).to(1).end()
                        .bind(Foo.class)
                            .annotatedWith(named("foo"))
                            .to(FooImpl.class)
                            .in(Singleton.class)
                        .end()
                        .bind(Bar.class).to(BarImpl.class).end()
                    .end()
                .build());
    }

    @Test
    public void testGuiceDSL() {
        assertInjector(Guice.createInjector(new AbstractModule() {
            @Override protected void configure() {
                bindConstant().annotatedWith(named("i")).to(1);
                bind(Foo.class)
                        .annotatedWith(named("foo"))
                        .to(FooImpl.class)
                        .in(Singleton.class);
                bind(Bar.class).to(BarImpl.class);
            }
        }));
    }

    private void assertInjector(final Injector injector) {
        final Bar bar1 = injector.getInstance(Bar.class);
        final Bar bar2 = injector.getInstance(Bar.class);
        assertNotSame(bar1, bar2);
        assertThat(bar1, is(instanceOf(BarImpl.class)));
        assertThat(bar2, is(instanceOf(BarImpl.class)));
        assertSame(bar1.foo, bar2.foo);
        assertThat(bar1.foo, is(instanceOf(FooImpl.class)));
        assertThat(bar1.foo.i, is(1));
    }

    private abstract static class Foo {

        int i;
    }

    private static class FooImpl extends Foo {

        @Inject
        FooImpl(@Named("i") int i) { this.i = i; }
    }

    private abstract static class Bar {

        Foo foo;
    }

    private static class BarImpl extends Bar {

        @Inject
        BarImpl(@Named("foo") Foo foo) { this.foo = foo; }
    }
}

File: neuron-di-guice/src/test/scala/global/tranquillity/neuron/di/guice/it/GuiceContextIT.scala
package global.tranquillity.neuron.di.guice.it

import javax.inject.{Inject, Named, Singleton}

import com.google.inject.{AbstractModule, Guice, Injector}
import global.tranquillity.neuron.di.guice.GuiceContext
import org.junit.runner.RunWith
import org.scalatest.Matchers._
import org.scalatest.WordSpec
import org.scalatest.junit.JUnitRunner
import com.google.inject.name.Names.named
import GuiceContextIT._

@RunWith(classOf[JUnitRunner])
class GuiceContextIT extends WordSpec {

  "A Guice context" should {
    "provide a nice DSL for building an injector" in {
      testInjector(
        new GuiceContext()
          .injector
            .module
              .bindConstant.annotatedWith(named("one")).to(1).end
              .bind(classOf[Foo])
                .annotatedWith(named("foo"))
                .to(classOf[FooImpl])
                .in(classOf[Singleton])
              .end
              .bind(classOf[Bar]).to(classOf[BarImpl]).end
            .end
          .build
      )
    }

    "build an injector with the same behavior as if using the original API" in {
      testInjector(
        Guice.createInjector(new AbstractModule {
          def configure() {
            bindConstant.annotatedWith(named("one")).to(1)
            bind(classOf[Foo])
              .annotatedWith(named("foo"))
              .to(classOf[FooImpl])
              .in(classOf[Singleton])
            bind(classOf[Bar]).to(classOf[BarImpl])
          }
        })
      )
    }
  }

  private def testInjector(injector: Injector) {
    injector getInstance classOf[Injector] should be theSameInstanceAs injector
    val bar1 = injector getInstance classOf[Bar]
    val bar2 = injector getInstance classOf[Bar]
    bar1 should not be theSameInstanceAs(bar2)
    bar1 shouldBe a[BarImpl]
    bar2 shouldBe a[BarImpl]
    bar1.foo shouldBe a[FooImpl]
    bar1.foo should be theSameInstanceAs bar2.foo
    bar1.foo.i shouldBe 1
  }
}

object GuiceContextIT {

  private trait Foo { def i: Int }

  private class FooImpl @Inject() (@Named("one") val i: Int) extends Foo

  private trait Bar { def foo: Foo }

  private class BarImpl @Inject() (@Named("foo") val foo: Foo) extends Bar
}
